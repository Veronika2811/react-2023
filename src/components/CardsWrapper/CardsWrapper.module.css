.main__characters {
  padding: 100px 20px;
}

.characters {
  display: flex;
  flex-wrap: wrap;
  justify-content: space-around;
  gap: 20px;
}