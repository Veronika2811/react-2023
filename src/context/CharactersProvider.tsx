import { PropsWithChildren, useState, useMemo } from 'react';
import { useSearchParams } from 'react-router-dom';

import { CharactersContext } from './context';
import {
  DEFAULT_VALUE_PER_PAGE,
  DETAILS_URL_PARAMETER_KEY,
  LOCAL_STORAGE_KEY,
} from '../constants/constants';
import { IData } from '../types/types';

export const CharactersProvider = ({ children }: PropsWithChildren) => {
  const [searchParams] = useSearchParams();

  const { Provider } = CharactersContext;

  const [searchQuery, setSearchQuery] = useState<string>(
    localStorage.getItem(LOCAL_STORAGE_KEY) || ''
  );
  const [perPage, setPerPage] = useState<number>(DEFAULT_VALUE_PER_PAGE);
  const [data, setData] = useState<IData | null>(null);
  const [detailedCard, setDetailedCard] = useState<string>(
    searchParams.get(DETAILS_URL_PARAMETER_KEY) || ''
  );

  const value = useMemo(
    () => ({
      searchQuery,
      setSearchQuery,
      perPage,
      setPerPage,
      data,
      setData,
      detailedCard,
      setDetailedCard,
    }),
    [searchQuery, perPage, data, detailedCard]
  );

  return <Provider value={value}>{children}</Provider>;
};
