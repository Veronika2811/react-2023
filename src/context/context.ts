import React, { createContext } from 'react';

import { IData } from '../types/types';

export type TCharactersContext = {
  searchQuery: string;
  setSearchQuery: React.Dispatch<React.SetStateAction<string>>;
  perPage: number;
  setPerPage: React.Dispatch<React.SetStateAction<number>>;
  data: IData | null;
  setData: React.Dispatch<React.SetStateAction<IData | null>>;
  detailedCard: string;
  setDetailedCard: React.Dispatch<React.SetStateAction<string>>;
};

export const CharactersContext = createContext<TCharactersContext>(null!);
