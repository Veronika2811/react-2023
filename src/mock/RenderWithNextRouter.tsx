import React, { ReactNode } from 'react';
import { RouterContext } from 'next/dist/shared/lib/router-context.shared-runtime';

import { createMockRouter } from './createMockRouter';

interface IRenderWithNextRouter {
  queryParams?: { [key: string]: string };
  children: ReactNode;
}

const RenderWithNextRouter = ({
  queryParams = {},
  children,
}: IRenderWithNextRouter) => {
  const { Provider } = RouterContext;

  return (
    <Provider value={createMockRouter({ query: queryParams })}>
      {children}
    </Provider>
  );
};

export default RenderWithNextRouter;
