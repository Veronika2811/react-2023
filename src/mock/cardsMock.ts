export const MortyMock = {
  created: '2017-11-04T18:50:21.651Z',
  episode: ['https://rickandmortyapi.com/api/episode/1'],
  gender: 'Male',
  id: 2,
  image: 'https://rickandmortyapi.com/api/character/avatar/2.jpeg',
  location: {
    name: 'Citadel of Ricks',
    url: 'https://rickandmortyapi.com/api/location/3',
  },
  name: 'Morty Smith',
  origin: {
    name: 'unknown',
    url: '',
  },
  species: 'Human',
  status: 'Alive',
  type: '',
  url: 'https://rickandmortyapi.com/api/character/2',
};

export const AlbertMock = {
  created: '2017-11-04T20:20:20.965Z',
  episode: ['https://rickandmortyapi.com/api/episode/12'],
  gender: 'Male',
  id: 11,
  image: 'https://rickandmortyapi.com/api/character/avatar/11.jpeg',
  location: {
    name: 'Earth (Replacement Dimension)',
    url: 'https://rickandmortyapi.com/api/location/20',
  },
  name: 'Albert Einstein',
  origin: {
    name: 'Earth (C-137)',
    url: 'https://rickandmortyapi.com/api/location/1',
  },
  species: 'Human',
  status: 'Dead',
  type: '',
  url: 'https://rickandmortyapi.com/api/character/11',
};
