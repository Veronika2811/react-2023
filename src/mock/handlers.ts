import { HttpResponse, http } from 'msw';

import { BASE_URL } from '@/utils/constants/constants';

export const handlers = [
  http.get(`${BASE_URL}`, () => {
    return HttpResponse.error();
  }),

  http.get(`${BASE_URL}/:id`, ({ params }) => {
    const { details } = params;
    return HttpResponse.json({
      created: '2017-11-04T18:50:21.651Z',
      episode: ['https://rickandmortyapi.com/api/episode/1'],
      gender: 'Male',
      id: details,
      image: 'https://rickandmortyapi.com/api/character/avatar/2.jpeg',
      location: {
        name: 'Citadel of Ricks',
        url: 'https://rickandmortyapi.com/api/location/3',
      },
      name: 'Morty Smith',
      origin: {
        name: 'unknown',
        url: '',
      },
      species: 'Human',
      status: 'Alive',
      type: '',
      url: 'https://rickandmortyapi.com/api/character/2',
    });
  }),
];