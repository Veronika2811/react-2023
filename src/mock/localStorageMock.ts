const mockLocalStorage = () => {
  const setItemMock = vi.fn();
  const getItemMock = vi.fn();

  beforeEach(() => {
    Storage.prototype.setItem = setItemMock;
    Storage.prototype.getItem = getItemMock;
  });

  afterEach(() => {
    setItemMock.mockRestore();
    getItemMock.mockRestore();
  });

  return { setItemMock, getItemMock };
};

export const { setItemMock, getItemMock } = mockLocalStorage();
