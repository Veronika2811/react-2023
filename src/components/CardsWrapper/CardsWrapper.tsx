import { useContext } from 'react';
import { useSearchParams } from 'react-router-dom';

import Card from '../Card/Card';
import NothingFound from '../NothingFound/NothingFound';
import changePerPage from '../../utils/changePerPage';
import { CharactersContext } from '../../context/context';
import { IDataResult } from '../../types/types';
import {
  ADDITIONAL_VALUE_PER_PAGE,
  DETAILS_URL_PARAMETER_KEY,
} from '../../constants/constants';

import classes from './CardsWrapper.module.css';

interface ICardsWrapperProps {
  cards: IDataResult[] | undefined;
  currentPage: number;
}

const CardsWrapper = ({ cards, currentPage }: ICardsWrapperProps) => {
  const [, setSearchParams] = useSearchParams();
  const { perPage } = useContext(CharactersContext);

  if (cards) {
    const characters =
      cards && perPage === ADDITIONAL_VALUE_PER_PAGE
        ? changePerPage(cards, currentPage, perPage)
        : cards;

    return (
      <ul
        className={`${classes.main__characters} ${classes.characters}`}
        onClick={(e) => {
          if (e.target === e.currentTarget) {
            setSearchParams((searchParams) => {
              searchParams.delete(DETAILS_URL_PARAMETER_KEY);
              return searchParams;
            });
          }
        }}
      >
        {characters.map((card) => (
          <Card card={card} key={card.id} />
        ))}
      </ul>
    );
  }

  return <NothingFound />;
};

export default CardsWrapper;
