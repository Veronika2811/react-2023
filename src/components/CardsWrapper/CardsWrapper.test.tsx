import { expect, it } from 'vitest';
import { HashRouter } from 'react-router-dom';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';

import { CharactersProvider } from '../../context/CharactersProvider';
import CardsWrapper from './CardsWrapper';
import { Albert_Mock, Morty_Mock } from '../../mock/cardsMock';

describe('CardsWrapper component', () => {
  it('renders correctly', () => {
    const container = render(
      <HashRouter>
        <CharactersProvider>
          <CardsWrapper cards={[Morty_Mock, Albert_Mock]} currentPage={1} />
        </CharactersProvider>
      </HashRouter>
    );
    expect(container).toMatchSnapshot();
  });

  it('should display the specified number of cards', () => {
    render(
      <HashRouter>
        <CharactersProvider>
          <CardsWrapper cards={[Morty_Mock, Albert_Mock]} currentPage={1} />
        </CharactersProvider>
      </HashRouter>
    );
    const element = screen.queryAllByTestId('card');
    expect(element).toHaveLength(2);
  });

  it('should be displayed component, when there are no cards', () => {
    render(
      <HashRouter>
        <CharactersProvider>
          <CardsWrapper cards={undefined} currentPage={1} />
        </CharactersProvider>
      </HashRouter>
    );

    const element = screen.queryByTestId('nothing-found');
    expect(element).toBeInTheDocument();
  });
});
