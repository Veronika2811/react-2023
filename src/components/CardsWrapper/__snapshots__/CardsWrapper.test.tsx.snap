// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CardsWrapper component > renders correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <ul
        class="_main__characters_4618fc _characters_4618fc"
      >
        <li
          class="_card_1f2933"
          data-testid="card"
        >
          <p
            class="_card__label_1f2933 _card__label_green_1f2933"
          >
            Alive
          </p>
          <img
            alt="Morty Smith"
            class="_card__image_1f2933"
            src="https://rickandmortyapi.com/api/character/avatar/2.jpeg"
          />
          <div
            class="_card__content_1f2933"
          >
            <h2>
              Morty Smith
            </h2>
            <p>
              Gender: Male
            </p>
            <p>
              Species: Human
            </p>
            <p>
              Location: Citadel of Ricks
            </p>
          </div>
        </li>
        <li
          class="_card_1f2933"
          data-testid="card"
        >
          <p
            class="_card__label_1f2933 _card__label_red_1f2933"
          >
            Dead
          </p>
          <img
            alt="Albert Einstein"
            class="_card__image_1f2933"
            src="https://rickandmortyapi.com/api/character/avatar/11.jpeg"
          />
          <div
            class="_card__content_1f2933"
          >
            <h2>
              Albert Einstein
            </h2>
            <p>
              Gender: Male
            </p>
            <p>
              Species: Human
            </p>
            <p>
              Location: Earth (Replacement Dimension)
            </p>
          </div>
        </li>
      </ul>
    </div>
  </body>,
  "container": <div>
    <ul
      class="_main__characters_4618fc _characters_4618fc"
    >
      <li
        class="_card_1f2933"
        data-testid="card"
      >
        <p
          class="_card__label_1f2933 _card__label_green_1f2933"
        >
          Alive
        </p>
        <img
          alt="Morty Smith"
          class="_card__image_1f2933"
          src="https://rickandmortyapi.com/api/character/avatar/2.jpeg"
        />
        <div
          class="_card__content_1f2933"
        >
          <h2>
            Morty Smith
          </h2>
          <p>
            Gender: Male
          </p>
          <p>
            Species: Human
          </p>
          <p>
            Location: Citadel of Ricks
          </p>
        </div>
      </li>
      <li
        class="_card_1f2933"
        data-testid="card"
      >
        <p
          class="_card__label_1f2933 _card__label_red_1f2933"
        >
          Dead
        </p>
        <img
          alt="Albert Einstein"
          class="_card__image_1f2933"
          src="https://rickandmortyapi.com/api/character/avatar/11.jpeg"
        />
        <div
          class="_card__content_1f2933"
        >
          <h2>
            Albert Einstein
          </h2>
          <p>
            Gender: Male
          </p>
          <p>
            Species: Human
          </p>
          <p>
            Location: Earth (Replacement Dimension)
          </p>
        </div>
      </li>
    </ul>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
